lib/Service/StackService.php

```diff
@@ -362,4 +368,93 @@ public function reorder($id, $order) {
 
 		return $result;
 	}
+
+	/**
+	 * @param $id
+	 * @param $boardId
+	 * @return Stack
+	 * @throws StatusException
+	 * @throws BadRequestException
+	 */
+	public function clone($id, $boardId) {
+		if (is_numeric($id) === false) {
+			throw new BadRequestException('stack id must be a number');
+		}
+
+		$this->permissionService->checkPermission(null, $boardId, Acl::PERMISSION_MANAGE);
+		if ($this->boardService->isArchived(null, $boardId)) {
+			throw new StatusException('Operation not allowed. This board is archived.');
+		}
+
+		$stack = $this->stackMapper->find($id);
+
+		$newStack = new Stack();
+		$newStack->setTitle($stack->getTitle() . ' (' . $this->l10n->t('copy') . ')');
+		$newStack->setBoardId($boardId);
+		$newStack->setOrder($stack->getOrder() +1);
+		$newStack = $this->stackMapper->insert($newStack);
+
+		$cards = $this->cardMapper->findAll($id);
+		foreach ($cards as $card) {
+			$newCard = new Card();
+			$newCard->setTitle($card->getTitle());
+			$newCard->setStackId($card->getStackId());
+			$newCard->setType($card->getType());
+			$newCard->setOrder($card->getOrder());
+			$newCard->setOwner($card->getOwner());
+			$newCard->setDescription($card->getDescription());
+			$newCard->setDuedate($card->getDuedate());
+			$newCard = $this->cardMapper->insert($newCard);
+
+			$this->activityManager->triggerEvent(ActivityManager::DECK_OBJECT_CARD, $card, ActivityManager::SUBJECT_CARD_CREATE);
+			$this->changeHelper->cardChanged($card->getId(), false);
+
+			$this->eventDispatcher->dispatch(
+				'\OCA\Deck\Card::onCreate',
+				new FTSEvent(
+					null, ['id' => $card->getId(), 'card' => $card, 'userId' => $owner, 'stackId' => $stackId]
+				)
+			);
+		}
+
+		
+
+
+		/*
+		$labels = $this->labelMapper->findAll($id);
+		foreach ($labels as $label) {
+			$newLabel = new Label();
+			$newLabel->setTitle($label->getTitle());
+			$newLabel->setColor($label->getColor());
+			$newLabel->setBoardId($newBoard->getId());
+			$this->labelMapper->insert($newLabel);
+		}
+
+		$stacks = $this->stackMapper->findAll($id);
+		foreach ($stacks as $stack) {
+			$newStack = new Stack();
+			$newStack->setTitle($stack->getTitle());
+			$newStack->setBoardId($newBoard->getId());
+			$this->stackMapper->insert($newStack);
+		}
+		*/
+
+
+
+		$this->activityManager->triggerEvent(
+			ActivityManager::DECK_OBJECT_BOARD, $newStack, ActivityManager::SUBJECT_STACK_CREATE
+		);
+		$this->changeHelper->boardChanged($boardId);
+
+		$this->eventDispatcher->dispatch(
+			'\OCA\Deck\Stack::onCreate',
+			new GenericEvent(null, ['id' => $newStack->getId(), 'stack' => $newStack])
+		);
```

Comment:
Please trigger this right after creating the stack and before the card creating so it is still emitted even if something goes wrong during card creation.